File: Programmers/Level2/가장큰정사각형찾기.py
from itertools import chain

def solution(board):
    answer = 0

    for i in range(1, len(board)):
        for j in range(1, len(board[0])):
            if board[i][j] == 1:
                board[i][j] = min(board[i-1][j-1], board[i-1][j], board[i][j-1]) + 1
    
    answer = max(list(chain(*board)))**2
    return answer

board = [[0,1,1,1],[1,1,1,1],[1,1,1,1],[0,0,1,0]]
print(solution(board))

File: Programmers/Level2/가장큰정사각형찾기/가장큰정사각형찾기.js
function solution(board) {
  let x = board[0].length;
  let y = board.length;
  let answer = 0;

  if (x < 2 || y < 2) return 1;

  for (let i = 1; i < y; i++) {
    for (let j = 1; j < x; j++) {
      if (board[i][j] > 0) {
        const min = Math.min(
          board[i - 1][j - 1],
          board[i][j - 1],
          board[i - 1][j]
        );
        board[i][j] = min + 1;
      }
      if (answer < board[i][j]) {
        answer = board[i][j];
      }
    }
  }
  return answer * answer;
}

const board = [
  [0, 1, 1, 1],
  [1, 1, 1, 1],
  [1, 1, 1, 1],
  [0, 0, 1, 0],
];

console.log(solution(board));
